### ISSUE TITLE
ssm: per-dimension observation mask for kalman_filter / parallel_kalman_filter

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/gauss_flows#139]
## Problem / Request

`kalman_filter` accepts `mask: Bool[Array, "T"]` — a **whole-step** gate that runs
predict-only. Partially-observed multivariate series need a **per-channel** gate:
at step `t`, some entries of `y_t ∈ ℝ^M` are observed and some are not. There is
currently no way to express that.

## User Story

> As a user modelling irregular multivariate time series, I want `kalman_filter`
> to accept a `(T, M)` observation mask, so that I can filter panels where
> different channels are measured at different times without padding, imputing,
> or splitting the series.

## Motivation

This unblocks the Normalizing Kalman Filter port (`gauss_flows` issues C/D) and is
independently useful for any `pyrox-gp` model over ragged multi-output data.

Two properties make it worth doing properly rather than leaving to callers:

1. The masked formulation is **exact** — it reproduces the row-deleted filter to
   machine precision, including the degenerate all-masked step, with no branching.
2. The naive implementation has a **silent likelihood bug** (see Mathematical
   Notes). It is a constant offset under a fixed mask, so it never shows up in
   development, and becomes a per-example bias the moment the mask varies — which
   is the only reason the feature exists.

Back-compat is non-negotiable: `pyrox_gp._markov` and
`pyrox_gp._inference_nongauss_markov` depend on the `(T,)` path for merged
train/test grids.

## Proposed API

```python
def kalman_filter(
    transition, obs_model, process_noise, obs_noise, observations,
    init_mean, init_cov,
    *,
    mask: Bool[Array, " T"] | Bool[Array, "T M"] | None = None,
    solver: AbstractSolverStrategy | None = None,
    woodbury_innovation: bool = False,
) -> FilterState:
    r"""...

    Args:
        mask: Optional observation mask.

            - Shape ``(T,)`` — per-step gate. ``False`` runs the predict step
              only and contributes nothing to the log-likelihood. Existing
              behaviour, unchanged.
            - Shape ``(T, M)`` — per-channel gate. ``False`` entries are
              marginalised out: the corresponding rows of ``H_t`` are zeroed,
              a unit block is substituted into ``R_t``, and the residual entry
              is set to zero. The returned ``log_likelihood`` is the exact
              marginal $\log p(y_{\mathrm{obs}})$, not the full-vector density.

            A ``(T, M)`` mask with an all-``False`` row is equivalent to a
            ``False`` entry in the ``(T,)`` form and needs no special-casing.
    """
```

Disambiguation is by `mask.ndim`, so no new keyword is needed. `M == 1` is
unambiguous either way (a `(T,)` mask and a `(T, 1)` mask coincide).

## Design Snapshot

The masked update, verified against a row-deleted reference (see Testing):

```python
mk = mask_t.astype(y_t.dtype)                       # (M,) in {0., 1.}
H_m = H_t * mk[:, None]                             # zero the unobserved rows
R_m = R_t * (mk[:, None] * mk[None, :]) + jnp.diag(1.0 - mk)
v   = (y_t - H_m @ x_pred) * mk                     # zero the unobserved residual
S   = H_m @ P_pred @ H_m.T + R_m                    # block-diagonal by construction

# ... standard gain / update, unchanged ...

# Likelihood: strip the spurious contribution of the dummy block.
n_off  = M - mk.sum()
ll_inc = -0.5 * (v @ Sinv_v + ld + M * _LOG_2PI) + 0.5 * n_off * _LOG_2PI
```

The existing update body at `src/gaussx/_ssm/_kalman.py:169-186` needs no change
beyond substituting `H_m`, `R_m`, `v` and adding the `n_off` term to `ll_inc`
(currently line ~185: `ll_inc = -0.5 * (v @ Sinv_v + ld + M * _LOG_2PI)`).

Why the posterior comes out exact without further work: with `H` row `i` zeroed,
`S` has zero off-diagonal coupling into channel `i` and `S[i,i] = 1`, so column
`i` of `K = P Hᵀ S⁻¹` is identically zero and the unobserved channel cannot move
the state.

## Mathematical Notes

```text
Model:   x_t = A_t x_{t-1} + q_t,   q_t ~ N(0, Q_t)
         y_t = H_t x_t     + r_t,   r_t ~ N(0, R_t)

Masked innovation covariance, with mask m ∈ {0,1}^M and dummy variance r_d = 1:

         S = H_m P H_mᵀ + R_m
           = [ S_obs   0 ]      (block-diagonal in the observed / masked split)
             [ 0     r_d·I ]

  log det S = log det S_obs + n_off · log r_d
  vᵀ S⁻¹ v  = v_obsᵀ S_obs⁻¹ v_obs            (masked residuals are zero)

So the naive per-step increment over-counts by

  −½ · n_off · (log 2π + log r_d)

i.e. exactly −½·n_off·log 2π ≈ −0.918939 per masked channel at r_d = 1.

The correction must be applied per step, NOT as a single constant at the end,
so that the result is invariant to r_d and to the mask pattern.
```

Two failure modes the correction must be tested against:

- **Silent under a fixed mask.** A constant offset does not change the argmax, so
  a fixed-mask test will pass with the bug present. Vary the mask across the
  batch.
- **Rewards complete series.** Without the correction, log-likelihood scales with
  the number of *missing* channels, so under a varying mask the model is
  gradient-pushed toward whichever series happen to be more complete.

## References & Existing Code

- Design doc: `kalman_survae_report.md` §2.1, §2.2, §4-A
- Reference impl (PyTorch, correct posterior / no likelihood): [`johannaSommer/KF_irreg_TS`](https://github.com/johannaSommer/KF_irreg_TS) `models/KF_base.py:204-240`
- Prior art for the row-deletion convention: Shumway & Stoffer, *Time Series Analysis and Its Applications*, §6.4 p. 311
- Existing `(T,)` mask path: `src/gaussx/_ssm/_kalman.py:51, 89, 121-127, 190-193`
- Parallel-scan mask path: `src/gaussx/_ssm/_parallel_kalman.py:59, 179, 200, 273-278`
- Downstream consumers that must not regress: `pyrox_gp/_markov.py:81-94`, `pyrox_gp/_inference_nongauss_markov.py:84, 172`

## Implementation Steps

- [ ] Extend `_normalise_tv_inputs` in `src/gaussx/_ssm/_utils.py:128` to accept and broadcast a `(T, M)` mask alongside the existing `(T,)` form
- [ ] Branch on `mask.ndim` in `kalman_filter` (`src/gaussx/_ssm/_kalman.py:42`): `ndim == 1` keeps the current `lax.cond` predict-only path; `ndim == 2` takes the masked-`H` path
- [ ] Apply the `+ 0.5 * n_off * _LOG_2PI` correction to `ll_inc`
- [ ] Mirror both in `parallel_kalman_filter` (`src/gaussx/_ssm/_parallel_kalman.py:170`) — the mask-aware "safe substitution" comment at line 278 already describes the pattern
- [ ] Confirm `rts_smoother` needs no change (it consumes filtered/predicted moments, which are mask-agnostic)
- [ ] Decide and document behaviour for `parallel_kalman_filter_sqrt` — either support the `(T, M)` mask or raise a clear `NotImplementedError`
- [ ] Update the `mask` docstring on both filters
- [ ] No new public symbols, so `docs/api/ssm.md` and `__init__.py` are unchanged — but re-read the `mask` prose in `docs/api/ssm.md` for accuracy

## Definition of Done

- [ ] Code lands at the intended path
- [ ] Public API exported via `src/gaussx/__init__.py` (n/a — signature change only)
- [ ] Tests pass: `make test`
- [ ] Lint + typecheck pass: `make lint && make typecheck`
- [ ] Docstrings (Google-style, `$...$` math, no Sphinx/RST) on all touched symbols

## Testing

New file `tests/ssm/test_kalman_channel_mask.py`. Measured targets below come from
`nkf_verify.py` §2–4 — treat them as numerical acceptance thresholds, not vibes.

- [ ] Unit test: filtered means match a row-deleted reference filter — **target ≤ 1e-14** (measured 2.22e-16)
- [ ] Unit test: filtered covariances match the row-deleted reference — **target ≤ 1e-14** (measured 1.11e-16)
- [ ] Unit test: `log_likelihood` matches the row-deleted reference — **target ≤ 1e-13** (measured 3.55e-15)
- [ ] **Regression test (the bug):** `log_likelihood` is invariant to the dummy variance. Parameterise over `r_dummy ∈ {1.0, 4.0, 100.0}`; all three must agree. Without the correction the measured values are −27.813393 / −36.131159 / −55.444414 against a true −16.786130
- [ ] Regression test: the naive-minus-corrected gap equals `n_missing × 0.5 × log(2π)` exactly — with 12 missing entries this is 11.027262
- [ ] Unit test: masked log-likelihood equals the exact marginal of the dense `(TM, TM)` joint Gaussian restricted to observed entries — **target ≤ 1e-13** (measured 3.55e-15)
- [ ] Edge case: a fully-unobserved step (`mask[t] = all False`) runs without NaN and leaves the state at its predicted value
- [ ] Edge case: `mask` all-`True` reproduces the unmasked filter bit-for-bit
- [ ] Back-compat: an existing `(T,)`-mask test is unchanged (use a genuinely partial mask, not all-`True`, or the test cannot catch a regression)
- [ ] Parity: `parallel_kalman_filter` agrees with `kalman_filter` under the same `(T, M)` mask
- [ ] float32: masked vs row-deleted relative log-likelihood difference ≤ 1e-5 at `T=60, M=6, Q=1e-4·I, R=1e-3·I` (measured 0.0, max mean error 2.38e-07)

## Documentation

- [ ] `docs/api/ssm.md` — extend the Kalman section prose to describe the two mask shapes
- [ ] Docstrings (covered by Definition of Done)

## Relationships

- Parent (theme epic): #
- Blocked by: #
- Blocks: # (gauss_flows: `MaskedLGSSM` base distribution)
- Related: # (gaussx: MFD discretisation)
